### PR TITLE
 BUG: Improve numerical precision/convergence failures of smirnov/kolmogorov

### PR DESCRIPTION
**Pull request by pvanmulbregt**
Addresses #7426, #7455, #7456, #7492.
scipy.special.cephes:
A rewrite of the algorithms in the Kolomgorov-Smirnov statistics package.
Kolmogorov: Improve efficiency of kolmogorov().
Fixed NANs, false convergence, unbounded computation in kolmogi().
Added dedicated functions (kolmogc/kolmogci/ kolmogp) to compute
the CDF, PPF and PDF.

Smirnov: Improve accuracy of smirnov().  Computation suffered from underflow
and/or denormalization, resulting in probabilities occasionally off by orders
of magnitude.  The SF/CDF is now usually within 1 ULP, the PDF within a few ULP.
Fixed NANs, convergence failures, false convergence in smirnovi().
Added dedicated functions (smirnovc/smirnovci/smirnovp) to compute
the CDF, PPF and PDF.
Added additional testing of K-S, removed xfail on previous tests.

scipy.stats: ksone and kstwobign classes.
Added dedicated _sf, _pdf, _isf methods to call new K-S functionality.
Updated _cdf and _ppf methods to make appropriate calls.

The accuracy improvements to smirnov() required increased precision
in some parts of the computation.  Ported most of the dd_real part of
the qd package and used in kolmogorov.c

### CODE REVIEW

**Comment by person142** [comment]
Off the top of my head--this PR is massive and looks like it could be split into several smaller PRs. Smaller PRs are much better for reviewers' sanity.

**Comment by pv** [comment]
A quick comment:

You cannot rely on the C compiler not optimizing `(a - b) + (b - c)` to 
`a - c` because IIUC the Intel compiler with its defaults settings will 
do that. Moreover, compilers may generate code for x86 FPU that use 
80-bit IEEE floats for intermediate computations.

Computations that rely on rounding intermediate results to float64 
should declare each such intermediate result as `volatile` to get better 
robustness vs. compiler unsafe math optimizations.

File: scipy/special/_legacy.pxd [hunk 1]
```diff
@@ -135,8 +135,20 @@ cdef inline double smirnov_unsafe(double n, double e) nogil:
     _legacy_cast_check("smirnov", n, 0)
     return smirnov(<int>n, e)
 
+cdef inline double smirnovc_unsafe(double n, double e) nogil:
```
Comment (person142) [inline scipy/special/_legacy.pxd]: No new functions should be added to `_legacy.pxd`; these exist only for back-compatibility reasons.

Comment (person142) [reply scipy/special/_legacy.pxd]: Ah never mind, being foolish.

Comment (pvanmulbregt) [reply scipy/special/_legacy.pxd]: When adding C code, my approach is a combination of reactive (fix what fails), and proactive (grep smirnov *.*; do same for smirnovc).  The changes to _legacy.pxd come from the latter.  Are you saying that I can ignore _legacy.pxd from now on when adding new functions to special?

Comment (person142) [reply scipy/special/_legacy.pxd]: Yes-the story is that in the past some Cephes functions accepted doubles and silently cast them to ints. The signatures were changed to fix this, but to preserve back-compatibility these wrappers were added which preserved the old behavior but added a warning. We don't want any new functions to have the implicit rounding, however.

**Comment by pvanmulbregt** [comment]
Q: How does one portably compile-time initialize a member of const struct with a NAN (or INFINITY)?

Attempt 1:
```const double2 DD_C_NAN = {{NPY_NAN, NPY_NAN}};```
"clang -arch x86_64 ..." complains:
scipy/special/cephes/dd_real.c:42:40: error: initializer element is not a compile-time constant

Attempt 2:
```const double2 DD_C_NAN = {{NAN, NAN}};```
"C:\Users\appveyor\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\amd64\cl.exe" complains:
 scipy\special\cephes\dd_real.c(42) : error C2065: 'NAN' : undeclared identifier

What is the right way to do this?

**Comment by pvanmulbregt** [comment]
Q: How to debug a travis build failure with missing symbol?

The 6th subtask of the travis build failed https://travis-ci.org/scipy/scipy/jobs/347966639
It fails due to a missing symbol `_two_split`.  `two_split()` happens to be a newly added inline function.  How can I get access to what the preprocessor and compiler are doing in this particular build?
Or is there some other recommended way to debug this? 

```
ImportError while importing test module '/Users/travis/build/scipy/scipy/build/testenv/lib/python3.6/site-packages/scipy/integrate/tests/test_banded_ode_solvers.py'.

Hint: make sure your test modules/packages have valid Python names.

Traceback:
scipy/integrate/__init__.py:89: in <module>
    from .quadrature import *
scipy/integrate/quadrature.py:10: in <module>
    from scipy.special import roots_legendre
scipy/special/__init__.py:646: in <module>
    from ._ufuncs import *
E   ImportError: dlopen(/Users/travis/build/scipy/scipy/build/testenv/lib/python3.6/site-packages/scipy/special/_ufuncs.cpython-36m-darwin.so, 2): Symbol not found: _two_split
E     Referenced from: /Users/travis/build/scipy/scipy/build/testenv/lib/python3.6/site-packages/scipy/special/_ufuncs.cpython-36m-darwin.so
E     Expected in: flat namespace
E    in /Users/travis/build/scipy/scipy/build/testenv/lib/python3.6/site-packages/scipy/special/_ufuncs.cpython-36m-darwin.so
```

**Comment by WarrenWeckesser** [comment]
I doubt I'll have time to take part in the review of this pull request, but I have one request:  if this pull request eventually gets approved, and if any *new* functions are created, be sure that their docstrings conform to the [docstring standard](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt).  I know there are many functions in `scipy.special` that don't follow the standard, but we shouldn't add any more.  (And be sure the new [docstrings have examples](https://github.com/scipy/scipy/issues/7168)!)

**Comment by pvanmulbregt** [comment]
Agreed. Will write conformant docstrings for the new functions, and bring the docstrings for the existing smirnov/smirnovc/kolmogorov/kolmogi up to current standard. Will also split this into two PRs, one for the double-double arithmetic, and one for the K-S improvements.

**Comment by pvanmulbregt** [comment]
Closing as splitting into multiple small PRs.

**Pull request closed without merging**